### ISSUE TITLE
optking convergence issues

### ISSUE THREAD

**Issue by ChayaSt** [psi4/psi4#647]
I was using `optking` to generate a torsion scan. Many of the structures generated didn't converge and even the ones that did converge didn't look quite right. Here is a script I used:

```python
#! Torsion scan of butane

molecule c4h10 {
  0 1
  H           -1.781487710695    -1.339015245276    -0.924843604480
  H           -2.809655081231    -0.316498366521     0.106834552942
  H           -1.696171739547    -1.492504703619     0.844458676121
  C           -1.808280241011    -0.768304160325     0.019321692974
  H           -0.781385143240     0.870036270346     0.998325840230
  H           -0.865561341910     1.021940295457    -0.757790563119
  C           -0.705914797210     0.292313565545     0.058091534575
  H            0.782112955956    -0.867612327944    -1.001664537213
  H            0.864488252487    -1.023755932239     0.754159895630
  C            0.705817366890    -0.292132845559    -0.060119676721
  H            1.697771204799     1.493890140835    -0.841405738393
  H            2.809791237804     0.316266152249    -0.104120355409
  H            1.780184254006     1.337108071239     0.927741879375
  C            1.808370373970     0.768135672436    -0.017435993935
}

# Now generate torsion scan 

steps = 36
start_angle = 0.001
step_size = 10

table = Table(rows=["D"], cols=["E(SCF)"])

for counter in range(0, steps):
    # set smaller basis for geometry optimization
    set{
        basis 6-31G*
        scf_type df
        guess sad
    }

    dihedral = 1.0 * counter * step_size + start_angle
    dihedral_string = "4 7 10 14 %.5f" % dihedral

    set optking{
        frozen_dihedral = ""
        fixed_dihedral = $dihedral_string
    }

    optimize('scf')  #geometry optimization with HF
    geom = c4h10.geometry()
    geom.print_out() # Print out optimized geometry

    # Print to screen
    print('-------------------------------------------')
    print('|   Optimized geometry for dihedral %.2f  |' % dihedral)
    print('-------------------------------------------')
    geom_py = mat2arr(geom)
    print(geom_py)

    for atom in range(c4h10.natom()):
        print("""{:4} {:4} {:12} {:8.4f} {:12.6f} {:12.6f} {:12.6f}   {:12.6f}""".format(
                                  c4h10.Z(atom),       # atomic number
                                  c4h10.symbol(atom),  # element symbol
                                  c4h10.label(atom),   # input element label
                                  c4h10.charge(atom),  # element charge
                                  c4h10.x(atom),       # x-coordinate
                                  c4h10.y(atom),       # y-coordinate
                                  c4h10.z(atom),       # z-coordinate
                                  c4h10.mass(atom),    # mass
              ))

    energy('scf') # HF single point energy
    escf = get_variable('SCF TOTAL ENERGY')
    table[dihedral] = [escf]

print('------------------------')
print('    Table of Energies   ')
print('------------------------')
print table
relative = table.copy()
relative.absolute_to_relative()
print('Relative table (Kcal/mol)')
print relative
```

I used a start angle of 0.001 instead of zero because when the angle was exactly 0 I got the following error:

```
Format of dihedral entry is "D atom_1 atom_2 atom_3 atom_4 (eq_val)"
Dynamic level is 0.
exc.g_really_quit() is 0.
```

**Comment by dgasmith** [comment]
A starting angle of 0 leads to an undefined dihedral. Can you explain what doesnt look right in the geometries that do converge?

**Comment by ChayaSt** [comment]
First of all, most structures don't converge. Only a few do. So that is an issue too.

When I generated my own structures to calculate the torsion drive, one half of the molecule was 
fixed while I rotated the other half (around the fixed central bond). In that case, most structures converged and when I compared those structures to a molecular mechanics energy (which were fit to reproduce a torsion drive at the same level of QM theory) the energies were very close.
When I used optiking, and looked at the structures, it seemed like both sides of the molecule were bending to achieve the fixed dihedral angle (not sure if I'm being clear here) and when I compared the energy of the structures that did converge to the molecular mechanics energy, the energies were higher by a factor of ~10.

**Comment by dgasmith** [comment]
@ChayaSt Thanks for the extra detail, that does seem strange. Perhaps @psi-rking can be of further assistance.

**Comment by andysim** [comment]
Sorry, busy week for me so I only just got a look at this problem.  I got the scan to work with a few changes.  First, your initial geometry has a dihedral of 180, but your initial step is 0, so that's an enormous perturbation to introduce through constraints; setting `start_angle = 181.0` sorts that out.  I tried 180, but the optimizer seems to have problems when it gets too close to 180.  I'm well out of my depth here, but I suspect [these](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/optking/set_params.cc#L719-L732) mysterious constants play a role.  What do you think, @psi-rking?  I removed the `frozen_dihedral` command, but I don't know if that makes a difference or not.  I also added `symmetry c1` to the molecule definition section, to make sure weird things don't happen as the symmetry changes.  Your `geom_py` command didn't work, so I changed it to `print(c4h10.save_string_xyz())`.

**Comment by ChayaSt** [comment]
> First, your initial geometry has a dihedral of 180, but your initial step is 0, so that's an enormous perturbation to introduce through constraints; setting start_angle = 181.0

>  I also added symmetry c1 to the molecule definition section, to make sure weird things don't happen as the symmetry changes.

I had done both of these things too (Sorry, I probably should've posted that too) and while the optimizations did converge (although not for all structures), the structures were still of very high energy when compared to the molecular mechanics energy of the optimized structure. When I generated the starting structures outside of psi4, the molecular mechanics energies were much closer to the QM energies (as it should be since the parameters were fit to reproduce this scan)

[Here](https://github.com/ChayaSt/torsionfit/blob/ddg/examples/butane/torsion_scans/psi4_archive/Psi4_QM_MM_comparison.pdf) is a figure comparing the QM and MM energies from structures generated with psi4
[Here](https://github.com/ChayaSt/torsionfit/blob/ddg/examples/butane/torsion_scans/psi4_archive/Pymol_scan_comparison.pdf) is one comparing QM and MM energies from structures generated outside of psi4.

**Comment by psi-rking** [comment]
Constrained optimizations are so often poorly (overly) interpreted and
erratic (dependent on initial guess, finite steps, ...) in their behavior
that I have never invested much time into them.  It would not surprise me
if there were a problem with them in PSI4.  However, I don't understand the
point of your first plot.  The HF energies of the structures are fine, but
the MM ones are way too high?  Or are these plots of different structures?

**Comment by loriab** [comment]
Sorry if this was tried early in your debugging, but have you looked at the geometries along the way? Varying just one torsion angle probably isn't going to work unless it's a very carefully constructed zmat. I had to vary several to get even [an ethane](https://github.com/loriab/psi4workshop/blob/master/index.ipynb) to work. Other programs have smarter guess-what-you-mean capabilities in this regard.

**Comment by andysim** [comment]
[This](https://github.com/psi4/psi4/files/866534/torsion.pdf) is the sort of geometry I get, which is very reasonable looking.  If you compare the 'good' and 'bad' geometries, is the different in the orientation of the methyl hydrogens?  If so, it's likely that your starting guess has them oriented with steric clashes, but the barrier to rotation is too high for the optimizer to be able to overcome.

**Comment by psi-rking** [comment]
Keep in mind that the optimizer (at least this one!) is not trying to
overcome any barriers except possibly one in the 'fixed coordinate'
direction.  It's not an IRC computation or a global search.  It's trying to
go downhill in all other directions, and it is subject to all the strange
behavior (discontinuities, etc.) present in scans by constrained
optimizations.  If you generate your own initial guesses to a constrained
optimization, you will get, in general, different results.

**Comment by ChayaSt** [comment]
> However, I don't understand the
point of your first plot.  The HF energies of the structures are fine, but
the MM ones are way too high?  Or are these plots of different structures?

No, these are plots of the same structure. That's why I'm not sure what is going on. 
>  If you generate your own initial guesses to a constrained
optimization, you will get, in general, different results.

Could it be that given these structures differ slightly from the ones that the forcefield was parameterized with that the MM energies are so much higher? In the second plot that I linked, I generated my own initial guesses and with those structures the MM energies are closer.
